### DISCUSSION TITLE
Change a variable without triggering a "panelupdate" event

### DISCUSSION THREAD

**Discussion by psheehymn** [gapitio/gapit-htmlgraphics-panel#159] [Q&A]
I have an HTML Graphics panel that contains several queries that I am processing in the JavaScript and combining into an HTML table. At the end of the process, I need to update several variables based on the table data that are used by other panels on the page. 

The table panel also needs to be aware of user changes to the dashboard time range, so the last line in my **onRender** section is to set a listener for the panel update:
`htmlNode.addEventListener('panelupdate', refreshTable);`

All of the work is being done in the **onRender** section (i.e. the **onInit** is empty) because the table needs to fully refresh on a time range change. The issue I am encountering is that the panel is also triggering on the variable updates it is doing itself and ends up continuously refreshing. I am setting the listener on the very last line, but it still seems to detect the preceding  variable changes retroactively.

Do I need to use a different type of listener or is there another work around to avoid the loop?

--- Addendum ---
After some additional experimentation, I am unsure if the "panelupdate" is being triggered by setting the variables, or the changes they trigger in other visualizations on the dashboard. I am leaning toward the latter as the cause and wonder if there is any way to work around that? Would moving the panel refresh to the **onInit** section and either or both of the _Dynamic htmlGraphics (onInit)_ or _Dynamic data object (onInit)_ options solve the issue?

**Comment by ZuperZee** [comment]
panelupdate and onRender are very similar as they trigger for the same reason.
The only reason to use panelupdate is to use it in onInit, which makes working with React, Solidjs, Svelte, etc.. a lot easier as you can have the data inside the "app". (https://github.com/gapitio/htmlgraphics-svelte-bundler-template/blob/5b99ffb59f95f8a7f913cbc0c2196be6c57be165/src/App.svelte#L10C59-L10C59)

Having panelupdate inside onRender would cause panelupdate event to be added to the eventlistener every time onRender runs.
Example code:
![panelupdate-in-onrender](https://github.com/gapitio/gapit-htmlgraphics-panel/assets/39485579/c89c9ab4-0351-4ed2-8ad0-3cc928eaccef)
`HTML`
```html
<div>
  <div>Render count: <span id="render-count" data-render-count="0">0</span></div>
  <div>PanelUpdate thing: <span id="thing" style="word-break: break-word;"></span></div>
</div>
```

`onRender`
```js
const rendersCountElt = htmlNode.querySelector("#render-count");
const thingElt = htmlNode.querySelector("#thing");

const renderCount = Number(rendersCountElt.textContent);
rendersCountElt.textContent = renderCount + 1;

console.log("ONRENDER");
thingElt.textContent += " ";
htmlNode.addEventListener("panelupdate", () => {
  thingElt.textContent += "!";
  console.log("PANELUPDATE");
});
```

`Panel options`
```json
{
    "calcsMutation": "standard",
    "reduceOptions": {
        "calcs": [
            "lastNotNull",
            "last",
            "firstNotNull",
            "first",
            "min",
            "max",
            "mean",
            "sum",
            "count",
            "range",
            "delta",
            "step",
            "diff",
            "logmin",
            "allIsZero",
            "allIsNull",
            "diffperc"
        ]
    },
    "add100Percentage": true,
    "centerAlignContent": true,
    "overflow": "visible",
    "useGrafanaScrollbar": true,
    "SVGBaseFix": true,
    "codeData": "{\n  \"text\": \"Random text\"\n}",
    "rootCSS": "",
    "css": "* {\n  font-family: Open Sans;\n}\n",
    "html": "<div>\n  <div>Render count: <span id=\"render-count\" data-render-count=\"0\">0</span></div>\n  <div>PanelUpdate thing: <span id=\"thing\" style=\"word-break: break-word;\"></span></div>\n</div>",
    "renderOnMount": true,
    "onRender": "const rendersCountElt = htmlNode.querySelector(\"#render-count\");\nconst thingElt = htmlNode.querySelector(\"#thing\");\n\nconst renderCount = Number(rendersCountElt.textContent);\nrendersCountElt.textContent = renderCount + 1;\n\nconsole.log(\"ONRENDER\");\nthingElt.textContent += \" \";\nhtmlNode.addEventListener(\"panelupdate\", () => {\n  thingElt.textContent += \"!\";\n  console.log(\"PANELUPDATE\");\n});\n",
    "panelupdateOnMount": true,
    "dynamicHtmlGraphics": false,
    "dynamicData": false,
    "dynamicFieldDisplayValues": false,
    "dynamicProps": false,
    "onInitOnResize": false,
    "onInit": ""
}
```

Updating/changing Grafana variables will trigger the panelupdate event and the onRender code. You can track the variables and check if they update. Here's a discussion about it: https://github.com/gapitio/gapit-htmlgraphics-panel/discussions/148.

If you're going to track Grafana variable changes I would have the code in onInit and use the panelupdate event. You could do it in onRender, but you would have to store the previous Grafana variable value in the dom/htmlNode or the global window element.

**Reply by psheehymn** [reply]
Thank you for the tip about the _panelupdate_ in the **onRender**; I had not realized that and have removed it. I'll review the rest of your answer more thoroughly and see if I can figure out how to solve my loop issue.

**Reply by psheehymn** [reply]
Please accept an advance apology, but it's not working for me. I duplicated my panel, moved the JavaScript to **OnInit** and set the _panelupdate_ listener in that section, per what I read in [#148](https://github.com/gapitio/gapit-htmlgraphics-panel/discussions/148). I have not added the variable tracking yet, I just wanted to get the new panel working with the changed code location.

I use a Custom variable on the dashboard that allows me to modify the values sent to the queries in the panel. When I change that variable, the old version of the panel (with the code in the **onRender**, minus the listener now) updates immediately (this has always worked, which is why I thought I needed everything in **onRender** before). However, the new version of the panel (with the code in **onInit** and including the _panelupdate_ listener) remains unchanged showing the prior data. I have the _Trigger panelupdate when mounted_, _Dynamic htmlGraphics (onInit)_, _Dynamic data object (onInit)_, _Dynamic fieldDisplayValues (onInit)_, and _Dynamic props (onInit)_ switches all set to On within the new panel. Interestingly, when I edit the new panel, the preview in the editor does show the new data.

Is there anything obvious I am missing that perhaps jumps out at you? I can send the JSON for the panels if that would help. 

Again, thank you for the assistance; I'm really _not_ trying to be an annoyance with the recent flurry of questions. I will also say that HTML Graphics is by far my favorite visualization, just because of all the flexibility it offers!

**Reply by ZuperZee** [reply]
I don't see an obvious flaw regarding the changes. If you could provide an example of what's happening, it would help a lot.

Below is a simple demo of the data updating with panelupdate in onInit.

Options changed:
`Dynamic data object (onInit)` to `true`

`HTML`
```html
<div style="text-align: center;">
  <div id="htmlgraphics-value">No data</div>
</div>
```

`onInit`
```js
function doSomething() {
  const valueElt = htmlNode.querySelector("#htmlgraphics-value");
  const value = data.series[0].fields[1].state.calcs.last;
  valueElt.textContent = value;
  console.log(value);
}

htmlNode.addEventListener("panelupdate", doSomething);
```

`Panel options`
```json
{
    "calcsMutation": "standard",
    "reduceOptions": {
        "calcs": [
            "lastNotNull",
            "last",
            "firstNotNull",
            "first",
            "min",
            "max",
            "mean",
            "sum",
            "count",
            "range",
            "delta",
            "step",
            "diff",
            "logmin",
            "allIsZero",
            "allIsNull",
            "diffperc"
        ]
    },
    "add100Percentage": true,
    "centerAlignContent": true,
    "overflow": "visible",
    "useGrafanaScrollbar": true,
    "SVGBaseFix": true,
    "codeData": "{\n  \"text\": \"Random text\"\n}",
    "rootCSS": "",
    "css": "* {\n  font-family: Open Sans;\n}\n",
    "html": "<div style=\"text-align: center;\">\n  <div id=\"htmlgraphics-value\">No data</div>\n</div>\n",
    "renderOnMount": true,
    "onRender": "",
    "panelupdateOnMount": true,
    "dynamicHtmlGraphics": false,
    "dynamicData": true,
    "dynamicFieldDisplayValues": false,
    "dynamicProps": false,
    "onInitOnResize": false,
    "onInit": "function doSomething() {\n  const valueElt = htmlNode.querySelector(\"#htmlgraphics-value\");\n  const value = data.series[0].fields[1].state.calcs.last;\n  valueElt.textContent = value;\n  console.log(value);\n}\n\nhtmlNode.addEventListener(\"panelupdate\", doSomething);\n"
}
```

`Panel JSON` (https://grafana.com/docs/grafana/latest/panels-visualizations/configure-panel-options/#view-a-panel-json-model)
```json
{
  "datasource": {
    "type": "grafana",
    "uid": "grafana"
  },
  "gridPos": {
    "h": 8,
    "w": 12,
    "x": 0,
    "y": 0
  },
  "id": 3,
  "options": {
    "calcsMutation": "standard",
    "reduceOptions": {
      "calcs": [
        "lastNotNull",
        "last",
        "firstNotNull",
        "first",
        "min",
        "max",
        "mean",
        "sum",
        "count",
        "range",
        "delta",
        "step",
        "diff",
        "logmin",
        "allIsZero",
        "allIsNull",
        "diffperc"
      ]
    },
    "add100Percentage": true,
    "centerAlignContent": true,
    "overflow": "visible",
    "useGrafanaScrollbar": true,
    "SVGBaseFix": true,
    "codeData": "{\n  \"text\": \"Random text\"\n}",
    "rootCSS": "",
    "css": "* {\n  font-family: Open Sans;\n}\n",
    "html": "<div style=\"text-align: center;\">\n  <div id=\"htmlgraphics-value\">No data</div>\n</div>\n",
    "renderOnMount": true,
    "onRender": "",
    "panelupdateOnMount": true,
    "dynamicHtmlGraphics": false,
    "dynamicData": true,
    "dynamicFieldDisplayValues": false,
    "dynamicProps": false,
    "onInitOnResize": false,
    "onInit": "function doSomething() {\n  const valueElt = htmlNode.querySelector(\"#htmlgraphics-value\");\n  const value = data.series[0].fields[1].state.calcs.last;\n  valueElt.textContent = value;\n  console.log(value);\n}\n\nhtmlNode.addEventListener(\"panelupdate\", doSomething);\n"
  },
  "targets": [
    {
      "datasource": {
        "type": "datasource",
        "uid": "grafana"
      },
      "refId": "A"
    }
  ],
  "title": "Panel Title",
  "type": "gapit-htmlgraphics-panel"
}
```

I don't mind the questions or find it annoying. It's great to hear you like the plugin :D

**Reply by psheehymn** [reply]
The attached screen capture demonstrates the issue. The lower panel (labeled "Old Version") is the panel with the code in **onRender**. the upper panel has it in **onInit** with the listener defined. I select a new value for the query and the old version updates, but the new one does not. Yet, if I go into edit on the new panel, it _did_ run the updated queries, but it did not refresh the panel shown on the dashboard.

https://github.com/gapitio/gapit-htmlgraphics-panel/assets/93273674/844be689-c26d-430a-9767-ffba60fe89bd

The code in both panels is identical other than which section it is in and the added listener in the **onInit** code.

**Reply by ZuperZee** [reply]
Can you create a minimal dashboard and share the code with the issue? I'm not able to reproduce it.

**Reply by psheehymn** [reply]
I've attached an importable export of a minimal dashboard that demonstrates the issue. Changing either the Journey dashboard variable or the time selector updates the onRender panel but not the onInit panel.
[onRender v onInit-1695855949240.json.txt](https://github.com/gapitio/gapit-htmlgraphics-panel/files/12743942/onRender.v.onInit-1695855949240.json.txt)

**Reply by ZuperZee** [reply]
The variables inside the onInit panel retrieves the grafana variables only at the first time the code runs.
If you move where you retrieve the variables to inside the refreshTable function.

The below code has `-> Section moved to here` and `Section moved from here` which are the changes.

`new onInit`
```js
// ********* //
// Functions //
// ********* //

// ------------------------------------
// Format time for display
// ------------------------------------
function formatTime(iTime) {
  //Time formatting options
  const options = { weekday: 'short', year: 'numeric', month: 'long', day: 'numeric', hour: "numeric", minute: "numeric", second: "numeric", timeZoneName: "short"};

  // Convert the input to a JS date/time object
  let jTime = new Date(0); // The 0 there is the key, which sets the date to the epoch
  jTime.setUTCMilliseconds(iTime);

  // Generate the time display string
  return jTime.toLocaleDateString(undefined, options);
}

// ----------------------------------------
// Retrieve the value of a Grafana variable
// ----------------------------------------
function getGrafanaVariableValue(variable) {
  return getTemplateSrv().replace(variable);
}

// ------------------------------------------------------------------------------------
// Update a grafana variable
// ------------------------------------------------------------------------------------
function updateGrafanaVariable(variableName, value) {
  htmlGraphics.locationService.partial(
    {
      [`var-${variableName}`]: value,
    },
    true // replace: true tells Grafana to update the current URL state, rather than creating a new history entry.
  );
}

// ------------------------------------
// Replace content of container section
// ------------------------------------
function replaceContentInContainer(id,content) {
  // Ignore errors from commneted out IDs
  try {
    // Get container
    let container = htmlNode.getElementById(id);

    // Update content
    container.innerHTML = content;
  } catch (err) {}
}

// ------------------------------------
// Refresh the table content
// ------------------------------------
function refreshTable() {

  // -> Section moved to here
  // ------------------------------------
  // Get the content of the Grafana variables
  let JourneyText = getGrafanaVariableValue(Journey).trim();
  let JourneyValue = getGrafanaVariableValue(Journeyv).trim();
  let FromTimeValue = getGrafanaVariableValue(FromTime).trim();
  let ToTimeValue = getGrafanaVariableValue(ToTime).trim();
  // ------------------------------------

  // The panel "remembers" the previous HTML dynamic row content on refresh, 
  //   so we need to reset the content back to just the original HTML
  
  // Declare flag comment string in original HTML
  const flag = "<!-- Dynamic Rows Start Here -->";

  // Truncate the table back to just the original HTML (retaining the flag string)
  htmlNode.innerHTML = htmlNode.innerHTML.slice(0,htmlNode.innerHTML.indexOf(flag) + flag.length);

  // Set Selected Time in the HTML
  replaceContentInContainer("selectedTime",formatTime(FromTimeValue)+" to<br>"+formatTime(ToTimeValue));

  // Set path name and list in the HTML
  replaceContentInContainer("pathname",JourneyText);
  replaceContentInContainer("pathlist",JourneyValue);

}

// ******************************************* //
// Begin Main Code                             //
// ******************************************* //
// Access the Custom Properties settings
const { Journey, Journeyv, FromTime, ToTime } = customProperties;

// -> Section moved from here
// ------------------------------------
// Get the content of the Grafana variables
// let JourneyText = getGrafanaVariableValue(Journey).trim();
// let JourneyValue = getGrafanaVariableValue(Journeyv).trim();
// let FromTimeValue = getGrafanaVariableValue(FromTime).trim();
// let ToTimeValue = getGrafanaVariableValue(ToTime).trim();
// ------------------------------------

// Declare the path list array
let paths = new Array();

// Do the initial build of the table
refreshTable();

// Refresh the table when the dashboard time range or variables change
htmlNode.addEventListener('panelupdate', refreshTable);

// Output HTML to developer console for debug (uncomment to troubleshoot)
//console.log(htmlNode.innerHTML);
```

**Comment by psheehymn** [comment]
Thank you, that was the trick. And after you pointed it out, it was blindingly obvious, I just couldn’t see it.

Now I need to morph the trick in #148 to skip running an update for variable changes this panel made, but process any others.

Thank you again!

Pat

**Reply by psheehymn** [reply]
Now that I have the panel properly monitoring for an update I care about, I have encountered a different but related issue. The queries in the panel take several seconds to execute, but the JavaScript appears to run almost immediately, so the data displayed often ends up being the prior dataset. Is there any way to get the panel to pause running the JavaScript until the queries have completed?

**Reply by ZuperZee** [reply]
In the data object, there is a `state` status.
![image](https://github.com/gapitio/gapit-htmlgraphics-panel/assets/39485579/4f811424-a757-4e38-8c3d-a322a0dc6eba)
These are the different states it can be
```ts
LoadingState {
    NotStarted = "NotStarted",
    Loading = "Loading",
    Streaming = "Streaming",
    Done = "Done",
    Error = "Error"
}
```

**Reply by psheehymn** [reply]
data.state is showing "Done" even as I can see the panel update still running (the sweeping stripe above the panel that replaced the old cycle icon). Could part of the issue with getting back the "old" data be because I am retrieving the query results by reference ID instead of sequentially?

`HitsQ = data.series.find(series => series.refId == "Hits");`

**Reply by ZuperZee** [reply]
Which datasource are you using?

**Reply by ZuperZee** [reply]
If you log the data in onRender and refresh, do you get two logs? One quickly and one after a bit once the query loads with state being "Done" on both?

`onRender`
```js
console.log(data)
```

**Reply by psheehymn** [reply]
Data source is Datadog. I actually got three logs because I misread your instructions and logged from both locations. First, one from onInit with a state of "Done" and then the predicted pair from onRender: the second at almost the same time asthe first (thousandths of a second later) with a state of "Loading" and then the third about .1 seconds later with state of "Done." The first two contain the prior dataset in the series values and the last on contains the expected new dataset.

**Reply by ZuperZee** [reply]
Is `Dynamic data object (onInit)` enabled?

**Reply by ZuperZee** [reply]
`Dynamic data object (onInit)` enabled
![slow-query-with-dynamic](https://github.com/gapitio/gapit-htmlgraphics-panel/assets/39485579/52b62df1-0170-486b-9a98-630155f039a3)
`Dynamic data object (onInit)` disabled
![slow-query-without-dynamic](https://github.com/gapitio/gapit-htmlgraphics-panel/assets/39485579/bdd24b7b-226a-48e6-979a-7a5130e3962c)

Code used in the gifs:
`onRender`
```js
console.log("OnRender", data);
```

`onInit`
```js
htmlNode.addEventListener("panelupdate", () => {
  console.log("OnInit", data);
});
```

**Reply by psheehymn** [reply]
I thought it was, but in checking, `Dynamic data object (onInit)` was not enabled. It is now and that does seem to have been the culprit; thank you! I've turned on all the onInit "when new data is available" switches now just in case.
